library/Pipeline/src/Juvix/Pipeline/Internal.hs

```diff
@@ -39,19 +43,77 @@ toCore paths = do
         Left errr -> pure $ Left (PipelineErr errr)
         Right con -> pure $ Right con
 
+extractTypeDeclar :: Context.Definition term ty a -> Maybe a
+extractTypeDeclar (Context.TypeDeclar t) = Just t
+extractTypeDeclar _ = Nothing
+
+mkDef ::
+  Sexp.T ->
+  Sexp.T ->
+  Context.SumT term1 ty1 ->
+  Context.T term2 ty2 Sexp.T ->
+  Maybe (Context.Def Sexp.T Sexp.T)
+mkDef typeCons dataConstructor s@Context.Sum {sumTDef, sumTName} c = do
+  t <- extractTypeDeclar . Context.extractValue =<< Context.lookup (NameSymbol.fromSymbol sumTName) c
+  declaration <- Sexp.findKey Sexp.car dataConstructor t
+  pTraceShowM ("Declaratione", typeCons, declaration, generateSumConsSexp typeCons declaration)
+  Just $
+    Context.D
+      { defUsage = Just Usage.Omega,
+        defMTy = generateSumConsSexp typeCons declaration,
+        defTerm = Sexp.list [Sexp.atom ":primitive", Sexp.atom "Builtin.Constructor"],
+        defPrecedence = Context.default'
+      }
+
+-- data Bar = Foo Bool Int
+-- (-> Bool (-> Int Bar)
+-- data Bar = Foo Bool Int F
+-- -> Bool -> Int -> F Bar
+-- data Bar = Foo Bool
+
+-- TODO: We're only handling the ADT case with no records or GADTs
+-- TODO: Handle records
+generateSumConsSexp :: Sexp.T -> Sexp.T -> Maybe Sexp.T
+generateSumConsSexp typeCons (Sexp.cdr -> declaration) = do
+  d <- sanitizeRecord <$> Sexp.foldr1 f declaration
+  pure $ Sexp.list [arrow, d, typeCons]
+  where
+    sanitizeRecord (x Sexp.:> fields)
+      | Sexp.isAtomNamed x ":record-d" = Sexp.list $ removeFieldNames fields
+    sanitizeRecord xs = xs
+    removeFieldNames fields
+      | Just l <- Sexp.toList (Sexp.groupBy2 fields) = g <$> l
+    g (Sexp.List [s, e]) = e
+    f n acc = Sexp.list [n, arrow, acc]
+    arrow = Sexp.atom "TopLevel.Prelude.->"
```

Comment:
There  are bits of specific structure that can be refactored out into the abstractions I've made in #767. Namely we should just run the record transformation and worry about that structure.